File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits: 
            return digits

        result = []
        table = {
            '2': 'abc', 
            '3': 'def', 
            '4': 'ghi', 
            '5': 'jkl', 
            '6': 'mno', 
            '7': 'pqrs', 
            '8': 'tuv', 
            '9': 'wxyz'
        }

        def dfs(nums: str, path: str):
            if len(digits) == len(path): 
                result.append(path)
                return 

            for i in  range(len(nums)):
                for char in table[nums[i]]: 
                    dfs(nums[i+1:], path + char)
        
        dfs(digits, '')
        return result

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<String>();
        if ( digits.length() == 0) {
            return result;
        }
        String[] mapping = new String[] {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        result.add("");

        for (int i = 0; i < digits.length() ; i++) {
            int x = digits.charAt(i) - '0';
            for (int j = 0; j < result.size() ; j++) {
                if(result.get(0).length() == i) {
                    String tmp2 = result.get(0);
                    result.remove(0);
                    for(String c : mapping[x].split("")) {
                             result.add(tmp2+c);
                    }
                }
            }
        }

        return result;
    }
}